### PR TITLE
Fix secondary, supplementary and unmapped duplicate reads removing

### PR DESCRIPTION
**Pull request by NataliaLeznikova**
**Description**

Initial issue:https://github.com/broadinstitute/picard/issues/451

According to an issue, "when the mapped mate of a singly mapping pair is marked as duplicate, its unmapped mate is not marked as duplicate. This is expected behaviour as MarkDuplicates ignores unmapped reads from consideration.
If, however,'REMOVE_DUPLICATES'=true, the singly mapped duplicates are removed but the unmapped mates remain in the resulting file. This is undesirable behaviour and these unmapped mates should also be removed."

**Implementation**

We added new collection 'nonProperlyMappedPairDuplicatesNames'. This collection is needed only if 'REMOVE DUPLICATES = true': it executes an additional pass through "doWork()" and collects names of secondary, supplementary and unmapped reads that are duplicates. At the final stage of writing new file it allows to check wether a secondary, supplementary or unmapped read is in duplicate set or not, and remove it, if it is.

We also added integration tests on the main functionality of MarkDuplicates and examined all metrics which are used by MarkDuplicates (DuplicationMetrics). Test cases are:

- marking duplicates with/without removing them (including optical ones);
- marking duplicates in files containing barcodes(tags BARCODE_TAG, READ_ONE_BARCODE_TAG, READ_TWO_BARCODE_TAG) with/without removing duplicates;
- files which include secondary or supplementary alignment duplicate reads;
- PicardException on the case: "This program requires input that are either coordinate or query sorted. Found unsorted".

**Performance**

We measured performance by benchmarking method runPicardCommandLine() with different arguments: different SAM/BAM files and checked both how program works with and without removing duplicates.

Comparing old and new implementations of MD (REMOVE_DUPLICATES = false)

|             MarkDuplicatesBenchmark                     |  Cnt  | Score (new impl.)| Score (old impl.) |Percent| Units| 
|:--------------------------------------------------------|:-----:|:----------------:|:-----------------:|:-----:|:----:|
| benchmark(big_dupes.sam) - 1000 reads                   |  50   |  1.364 ± 0.084   |   1.198 ± 0.080   | +14%  |ops/ms|

Comparing old and new implementations of MD (REMOVE_DUPLICATES = true)

|             MarkDuplicatesBenchmark                        |  Cnt  | Score (new impl.)| Score (old impl.) |Percent| Units| 
|:----------------------------------------------------------|:-----:|:----------------:|:-----------------:|:-----:|:----:|
| benchmark(6747_snippet_markduplicates.bam) - 276000 reads |  50   |  0.131 ± 0.014   |   0.110 ± 0.007   |  +19% |ops/ms|
| benchmark(big_dupes.sam) - 1000 reads                     |  50   |  1.049 ± 0.100   |   0.880 ± 0.063   |  +19% |ops/ms|

### CODE REVIEW

File: src/main/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 1]
```diff
@@ -935,5 +955,42 @@ public int compare(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDup
...
 
+        final CloseableIterator<SAMRecord> iteratorForRemoveDuplicates = headerAndIteratorForRemoveDuplicates.iterator;
+        while (iteratorForRemoveDuplicates.hasNext()) {
+            final SAMRecord rec = iteratorForRemoveDuplicates.next();
+
+            final boolean nextDuplicateIndexIsNecessary = recordInFileIndex > nextDuplicateIndex &&
+                    (sortOrder == SAMFileHeader.SortOrder.coordinate || !rec.getReadName().equals(duplicateQueryName));
+
+            if (nextDuplicateIndexIsNecessary) {
+                nextDuplicateIndex = (duplicateIndexes.hasNext() ? duplicateIndexes.next() : NO_SUCH_INDEX);
```
Comment (tlangs) [inline src/main/java/picard/sam/markduplicates/MarkDuplicates.java]: Input variables shouldn't be mutated. A new variable should be introduced to hold this value.

File: src/main/java/picard/sam/markduplicates/MarkDuplicates.java [hunk 2]
```diff
@@ -656,6 +670,8 @@ private ReadEndsForMarkDuplicates buildReadEnds(final SAMFileHeader header, fina
      */
     private void generateDuplicateIndexes(final boolean useBarcodes, final boolean indexOpticalDuplicates) {
         final int entryOverhead;
+
+
```
Comment (tlangs) [inline src/main/java/picard/sam/markduplicates/MarkDuplicates.java]: Unnecessary whitespace

File: src/test/java/picard/sam/markduplicates/MarkDuplicatesExceptionTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import picard.PicardException;
+import picard.cmdline.CommandLineProgramTest;
+
+import java.io.File;
+import java.util.*;
+
+public class MarkDuplicatesExceptionTest extends CommandLineProgramTest {
```
Comment (tlangs) [inline src/test/java/picard/sam/markduplicates/MarkDuplicatesExceptionTest.java]: The tests in this class could go in `MarkDuplicatesIntegrationTests`

File: src/test/java/picard/sam/markduplicates/MarkDuplicatesIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        args.add("OUTPUT=" + output.getAbsolutePath());
+        final File metrics = new File(outputDir, TEST_BASE_NAME + ".integration_metrics");
+        args.add("METRICS_FILE=" + metrics.getAbsolutePath());
+
+        Assert.assertEquals(runPicardCommandLine(args), 0);
+        examineDuplicationMetrics(input, output, metrics, false);
+    }
+
+    @Test(dataProvider = "MarkDuplicatesIntegrationTest")
+    public void markDuplicatesWithRemovingDuplicatesIntegrationTest(final String input) {
```
Comment (tlangs) [inline src/test/java/picard/sam/markduplicates/MarkDuplicatesIntegrationTests.java]: `markDuplicatesWithRemovingDuplicatesIntegrationTest` and `secondarySupplementaryUnmappedWithRemovingDuplicatesIntegrationTest` are exactly the same. The contents of the test should be pulled out into its own test and the two separate data providers should be one line methods that call the abstracted test.

**Comment by yfarjoun** [comment]
I'm not happy with the fact that the implementation includes a **third** traversal over the input file. I think that the stated result (removing duplicate for reads whose mate is unmapped) without resorting to a third pass.

here are my thoughts:

1. When a read is marked as a duplicate, if it's mate is unmapped, store it's name in a DiskBackedSortingCollection<String>. 
2. When writing the resulting file to disk, filter out reads that are unmapped && whose name is in the collection.

Thoughts?

**Comment by NataliaLeznikova** [comment]
Hello, @yfarjoun!

Thank you for your answer. 

As far as we have understood, we don't have information about reads' mates at the time of the second traversal. We suppose that it's impossible to find out whether a duplicate read has a mate or not as mates can be located at the different parts of the file, so that is why we are offering this implementation. 

Please, correct us if we are wrong.

**Comment by yfarjoun** [comment]
I think that unmapped reads whos mates are marked as duplicates, should also be marked as duplicates....if this isn't happening now, I think it's a bug...I'll open an issue about it.

**Comment by NataliaLeznikova** [comment]
Hello, @yfarjoun!

We definitely agree with you that there is no sense in diving into mark duplicates algorithm until we are not sure if it's actually a bug. We measured the performance on 17.35GB bam and figured out that after the 'fixed' branch worked twice longer than master one.

So there is one thing that seems to be obvious - the third traversal through file leads to performance decrease. We will be waiting for your decision on this point.

**Comment by yfarjoun** [comment]
no one objected to my suggestion that we close this issue...so I'm following through.
